app/assets/javascripts/dashboard/dashboard.js

```diff
@@ -443,39 +448,64 @@ MinionPoller = {
     var masterChecked = '';
     var minionHtml;
     var minionChecked = '';
+    var roleHtml = '';
+    var actionHtml = '';
 
-    if (isMaster) {
-      masterChecked = 'checked';
-    }
+    switch (minion.highstate) {
+      case "not_applied":
+        if (isMaster) {
+          masterChecked = 'checked';
+        }
 
-    masterHtml = '\
-      <input name="roles[master][]" id="roles_master_' + minion.id +
-      '" value="' + minion.id + '" type="checkbox" title="Select node as master" ' + masterChecked + '>';
+        masterHtml = '\
+          <input name="roles[master][]" id="roles_master_' + minion.id +
+          '" value="' + minion.id + '" type="checkbox" title="Select node as master" ' + masterChecked + '>';
 
-    if (isWorker) {
-      minionChecked = 'checked';
-    }
+        if (isWorker) {
+          minionChecked = 'checked';
+        }
 
-    minionHtml = '<input name="roles[worker][]" id="roles_worker_' + minion.id +
-      '" value="' + minion.id + '" type="checkbox" title="Select node for bootstrapping" ' + minionChecked + '>';
-
-    var roleHtml = '\
-      <td class="role-column">' +
-        masterHtml +
-        minionHtml +
-        '<div class="btn-group role-btn-group">\
-          <button type="button" class="btn btn-default master-btn '+ (isMaster ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '" data-minion-role="master">Master</button>\
-          <button type="button" class="btn btn-default worker-btn ' + (isWorker ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '" data-minion-role="worker">Worker</button>\
-          <button type="button" class="btn btn-default unused-btn ' + (isUnused ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '">Unused</button>\
-        </div>\
-      </td>\
-    ';
+        minionHtml = '<input name="roles[worker][]" id="roles_worker_' + minion.id +
+          '" value="' + minion.id + '" type="checkbox" title="Select node for bootstrapping" ' + minionChecked + '>';
+        roleHtml = '\
+          <td class="role-column" id="role-column-'+ minion.minion_id +'">' +
+            masterHtml +
+            minionHtml +
+            '<div class="btn-group role-btn-group">\
+              <button type="button" class="btn btn-default master-btn '+ (isMaster ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '" data-minion-role="master">Master</button>\
+              <button type="button" class="btn btn-default worker-btn ' + (isWorker ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '" data-minion-role="worker">Worker</button>\
+              <button type="button" class="btn btn-default unused-btn ' + (isUnused ? 'btn-primary' : '') + '" data-minion-id="' + minion.id + '">Unused</button>\
+            </div>\
+          </td>\
+        ';
+        actionHtml = '\
```

Comment:
You can move everything from above this line out of the switch statement and for the buttons you can add something like

```js
var rolesDisabled = State.pendingRemovalMinionId != null;
// (...)

// add disabled attribute to the buttons like
// ...
'<button (...) disabled="' + rolesDisabled + '">Master</button>'\
'<button (...) disabled="' + rolesDisabled + '">Worker</button>'\
'<button (...) disabled="' + rolesDisabled + '">Unused</button>'\
// ...
```